docs/tutorials/lognamespacing.md

```diff
@@ -0,0 +1,229 @@
+    async fn build(&self, cx: SourceContext) -> Result<super::Source> {
+        let log_namespace = cx.log_namespace(self.log_namespace);
+```
+
+The `cx.log_namespace` function gives us a `LogNamespace` enum that we can use to
+set the fields in the appropriate section of the Event.
+
+This `log_namespace` variable needs to be passed to any functions that will insert 
+data into the log event that is emitted by the source.
+
+### Vector metadata
+
+The Vector namespace contains data pertinent to how the event was ingested into 
+Vector. Currently two fields need to be added to this namespace - `ingest_timestamp`
+and `source_type`:
+
+```rust
+  self.log_namespace.insert_vector_metadata(
+      &mut log_event,
+      path!(self.timestamp_key()),
+      path!("ingest_timestamp"),
+      chrono::Utc::now(),
+  );
+            
+  self.log_namespace.insert_vector_metadata(
+      &mut log_event,
+      path!(self.source_type_key()),
+      path!("source_type"),
+      DnstapConfig::NAME,
+  );
+```
+
+If we look at the parameters to `insert_vector_metadata`:
+
+```rust
+    pub fn insert_vector_metadata<'a>(
+        &self,
+        log: &mut LogEvent,
+        legacy_key: impl ValuePath<'a>,
+        metadata_key: impl ValuePath<'a>,
+        value: impl Into<Value>,
+    ) 
+```
+
+#### log
+
+This needs to be the log event that is being populated.
+
+#### legacy_key
+
+This is the name of the field the timestamp is to be inserted into
+when using the Legacy Namespace. This will typically be values returned
+by calls to `log_schema().source_type_key()` and `log_schema().timestamp_key()`.
+
+#### metdata_key
+
+The name of the field when it is inserted into the Vector namespace. This 
+will be `path!("ingest_timestamp")` or `path!(source_type)`. The field names
+can be hardcoded since they are going into the Vector namespace, so conflicts
+with other field names cannot occur.
+
+#### value
+
+The actual value to be placed into the field. 
+
+For the ingest timestamp this will be `chrono::Utc::now()`. Source type will be
+the `NAME` property of the `Config` struct. `NAME` is provided by the 
+`configurable_component` macro. You may need to include `use vector_config::NamedComponent;`.
+
+### insert_standard_vector_source_metadata
+
+A utility function has been provided that can be used in a lot of cases to 
+insert both these fields into the Vector namespace:
+
+```rust
+
+  log_namespace.insert_standard_vector_source_metadata(
+      log,
+      KafkaSourceConfig::NAME,
+      Utc::now(),
+  );
+```
+
+### Source Metadata
+
+Other fields that describe the event - but are not the actual data for the event
+should go into the source metadata. Examples of source metadata are 
+
+- The Kafka topic when pulling from a Kafka stream.
+- Severity and Facility fields from a Syslog message.
+- The file path when pulling data from a file.
+- The timestamp extracted from the message (this can be different to the time
+  the message was ingested.)
```

Comment:
🤔 wouldn't this be actual event data (at the `.` path), while the source metadata would be (`kubernetes_logs` example) when the container runtime processed the log emitted by the application(s).

Reply:
Ah , yes.